File: Cpp/3714_longest_balanced_substring_ii.cpp
// https://leetcode.cn/problems/longest-balanced-substring-ii/

#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int longestBalanced(string s) {
        auto longestBalancedOneChar = [&]() {
            int res = 0;
            int n = s.size(), left = 0;
            while (left < n) {
                int right = left;
                while (right < n && s[right] == s[left]) {
                    right += 1;
                }
                res = max(res, right - left);
                left = right;
            }
            return res;
        };

        auto longestBalancedTwoChars = [&](const char ch1, const char ch2) {
            int res = 0;
            int n = s.size(), i = 0;
            while (i < n) {
                int diff = 0;
                unordered_map<int, int> record {{diff, i-1}};
                while (i < n && (s[i] == ch1 || s[i] == ch2)) {
                    if (s[i] == ch1) diff += 1;
                    else diff -= 1;
                    if (record.count(diff)) {
                        res = max(res, i - record[diff]);
                    }
                    else {
                        record[diff] = i;
                    }
                    i += 1;
                }
                i += 1;
            }
            return res;
        };

        auto longestBalancedThreeChars = [&]() {
            int res = 0;
            int n = s.size();
            // Convert (diff_1, diff_2) to (diff_1 + n) << 32 | (diff_2 + n).
            long long diff = ((long long)n << 32) | n;
            unordered_map<long long, int> record {{diff, -1}};
            unordered_map<char, int> cnt;
            for (int i = 0; i < n; ++i) {
                cnt[s[i]] += 1;
                int diff_1 = cnt['a'] - cnt['b'], diff_2 = cnt['a'] - cnt['c'];
                diff = ((long long)(diff_1 + n) << 32) | (diff_2 + n);
                if (record.count(diff)) {
                    res = max(res, i - record[diff]);
                }
                else {
                    record[diff] = i;
                }
            }
            return res;
        };

        int res = longestBalancedOneChar();
        res = max(res, longestBalancedTwoChars('a', 'b'));
        res = max(res, longestBalancedTwoChars('a', 'c'));
        res = max(res, longestBalancedTwoChars('b', 'c'));
        res = max(res, longestBalancedThreeChars());
        return res;
    }
};

int main() {
    Solution sol;
    // string s = "abbac";
    string s = "abc";
    cout << sol.longestBalanced(s) << '\n';
    return 0;
}

File: Python3/3714_longest_balanced_substring_ii.py
# https://leetcode.cn/problems/longest-balanced-substring-ii/

class Solution:
    def longestBalanced(self, s: str) -> int:
        # The substring contains characters of one type.
        res = self.longestBalancedOneChar(s)
        
        # The substring contains characters of two type.
        res = max(res, self.longestBalancedTwoChars(s, "a", "b"))
        res = max(res, self.longestBalancedTwoChars(s, "a", "c"))
        res = max(res, self.longestBalancedTwoChars(s, "b", "c"))

        # The substring contains characters of three type.
        res = max(res, self.longestBalancedThreeChars(s))
        
        return res
    
    def longestBalancedOneChar(self, s: str):
        res = 0
        n, left = len(s), 0
        while left < n:
            right = left
            while right < n and s[right] == s[left]:
                right += 1
            res = max(res, right-left)
            left = right
        return res
    
    def longestBalancedTwoChars(self, s: str, ch1: str, ch2: str):
        res = 0
        n, i = len(s), 0
        while i < n:
            diff = 0
            record = {diff: i-1}
            while i < n and (s[i] == ch1 or s[i] == ch2):
                diff += 1 if s[i] == ch1 else -1
                if diff in record:
                    res = max(res, i - record[diff])
                else:
                    record[diff] = i
                i += 1
            i += 1
        return res
    
    def longestBalancedThreeChars(self, s: str):
        res = 0
        record = {(0, 0): -1}
        cnt = {"a": 0, "b": 0, "c": 0}
        for i, ch in enumerate(s):
            cnt[ch] += 1
            diff = (cnt["a"]-cnt["b"], cnt["a"]-cnt["c"])
            if diff in record:
                res = max(res, i - record[diff])
            else:
                record[diff] = i
        return res